### ISSUE TITLE
IP Filtering Clarification

### ISSUE THREAD

**Issue by JL421** [storj/storj#2391]
After reading the blog: https://storj.io/blog/2019/06/ip-filtering-keeps-data-distributed/ and being a network engineer by trade I have serious concerns about the methods used to "ensure distributed data". I have raised the question in RocketChat but it has been ignored, so I'm raising it as an issue, since improper implementation of this function will have consequences that do not seem to have been considered.

Specifically this quote will cause the most issue: 

> With today’s v0.14.3 release, we’ve implemented a feature called IP filtering, which will ensure that no file pieces corresponding to the same file are stored in the same geographical area, based on logical subnets. 

Without knowing specifics about how these subnets are determined we have to assume something simple like using a /24 block subnet by default:

* Multiple ISPs can service one geographic region
  * By filtering a subnet you are only restricting a specific block served by a specific ISP.
  * This doesn't actually ensure geographically distributed data.
* An ISP may implement Carrier Grade NAT, allowing them to conserve public address space and cover a large customer base with a small subnet.
  * This means that the subnet filtered may actually service a very broad geographic area, also failing the stated goals.
* An IP does not tie back to a specific location.
  * With the invent of technologies like [LISP](https://en.wikipedia.org/wiki/Locator/Identifier_Separation_Protocol), you are literally separating a location from an identifier (IP Address).
  * Carriers are starting to implement these sorts of technologies to help combat IPv4 starvation and conserve their public address space.
* It actually pits SNOs on an ISP against each other and can cause the inverse effect of **more** concentrated data in a single location
  * SNO1 and SNO2 are grouped in the same subnet
  * SNO1 has 5 nodes
  * SNO2 has 1 node
  * By the grouping SNO1 will store more data because they will end up receiving 5 out of every 6 pieces sent to the group.
* It gives large data centers an advantage.
  * By having multiple carriers available in a facility, with numerous discontinuous IP blocks available, a data center could end up having a majority of stored customer data. Especially as these nodes would be higher performing with more bandwidth available.

If the filter is built such that you are actually looking at the ISP IP block allocations, or advertised prefixes :

* ISP block allocations may actually end up being split across vast geographic regions as customer demand required.
* ISP Prefix Advertisements will often be summarized (aggregated) to reduce the number of routes in the Global Internet routing table. Summarizing your routes is just being a good netizen.
Both of these scenarios, at best could aggregate much larger than anticipated sets of SNOs together and reduce network performance.

-----

How do you solve the distribution problem, without encountering these issues? In short, you don't. In long:
* Stop over-engineering solutions to try to prevent data clustering. Bluntly, your methods for doing so are flawed. At best, they can exclude large geographic areas from normal selection, at worst actually lead to increased centralization.
* Keep IP grouping simple, a single IP can include N nodes, but have data distributed as if they were a single node.
* Let the already described reputation system handle SNOs that are trying to game the system. If they are overtaxing their available resources they will be naturally limited as a result of having under-performing nodes.
* In the node selection process introduce randomness. Randomly select nodes that fit the customer's requirements (Geofenced, reputation, etc.), and not just the best.
* Accept that true, enforced decentralization, is not easily code-able, but rather involves human curation and verification. If customers want this type of service, offer it as a tier to offset the work required to verify and track SNOs. Pay verified SNOs a premium to store this enforced data. Build a database of actual locations that can be queried and served as requested.

-----

Final note: I request that someone from the core Storj Labs team to actually read this. Don't just paste generic responses boiling down to "decentralization at all costs". Don't just close it saying "We're working on it." If there isn't an answer, leave it open until there is. If you can't publicly disclose this information, that's fine, just give some reassurance that this has actually been thought about and resolved.

**Comment by littleskunk** [comment]
In storj v2 we had thousends of nodes running on the same IP. We noticed that the majority of the operators only compare the short term profit. If 2 storage nodes on the same IP get more data than a single storage node they will spin up thousends of them. In fact they started doing that in v3 as well. Sure we will disqualify them soon or later but at that point the damage is already done. We want to incentivize to run less but healty nodes from the beginning.

We are aware that this solution is not perfect.

**Comment by thepaul** [comment]
> After reading the blog: https://storj.io/blog/2019/06/ip-filtering-keeps-data-distributed/ and being a network engineer by trade I have serious concerns about the methods used to "ensure distributed data". I have raised the question in RocketChat but it has been ignored, so I'm raising it as an issue, since improper implementation of this function will have consequences that do not seem to have been considered.

Thanks for bringing it up. Your concerns are certainly valid ones. This method of clustering avoidance is indeed flawed, and we understand that, but so far we've determined that even a simplistic and crude heuristic like this yields more benefit than it does harm. It's important for us to stipulate that this approach is not a complete answer to data clustering. It suffers from both false positives and false negatives, but the rates of both are expected to be fairly low, and the existence of these cases does not break the approach.

> > With today’s v0.14.3 release, we’ve implemented a feature called IP filtering, which will ensure that no file pieces corresponding to the same file are stored in the same geographical area, based on logical subnets.

This quote might not have used the best wording. Of course it won't "ensure" that no file pieces are stored in the same geographical area (however you might define it); as you point out, that would be quite impossible.

> Without knowing specifics about how these subnets are determined we have to assume something simple like using a /24 block subnet by default:

That is indeed the current heuristic; a /24 prefix for ipv4 addresses and a /64 prefix for ipv6.

> * Multiple ISPs can service one geographic region
>   * By filtering a subnet you are only restricting a specific block served by a specific ISP.
>   * This doesn't actually ensure geographically distributed data.

Certainly accurate. By our experience across many common ISPs worldwide, however, it is a reasonably good approximator.

> * An ISP may implement Carrier Grade NAT, allowing them to conserve public address space and cover a large customer base with a small subnet.
>   * This means that the subnet filtered may actually service a very broad geographic area, also failing the stated goals.

Our research indicates:

* CGN is not widely deployed outside of cellular networks, and we don't think many SNOs will run nodes on cellular networks. (This needs some followup though; in how many markets would it be preferable to use a cellular provider over a non-cellular provider, and would it even be feasible to run the SN software there, given high connection counts and throughput/uptime requirements?)
* In most cases where CGN is deployed for non-cellular networks, users are also given a publicly routable IPv6 /64 subnet
* In most cases where CGN is deployed for non-cellular networks and where users are _not_ given IPv6 addresses, users are also not able to set up static inbound port mapping, so it would not be possible to run a storage node anyway.

> * An IP does not tie back to a specific location.
>   * With the invent of technologies like LISP, you are literally separating a location from an identifier (IP Address).

I was not aware of LISP. Thanks for the pointer! It looks like LISP adoption would actually make this approach more effective, as one of the guiding principles is that "RLOCs must be allocated in a way that is congruent with the topology of the network". Thus sites with a common RLOC (or possibly we'd look at a common RLOC prefix) will be _more_ likely to be located in the same physical space, as they are guaranteed to be in the same topological spot on the network. Have I got that wrong?

>   * Carriers are starting to implement these sorts of technologies to help combat IPv4 starvation and conserve their public address space.

Certainly true. It does not appear, though, that any ISPs _require_ all traffic to be encapsulated with LISP, and the Storj network is unlikely to make use of LISP encapsulation either (since it establishes endpoint identity in a different way), so LISP will probably not be a major factor here.

> * It actually pits SNOs on an ISP against each other and can cause the inverse effect of more concentrated data in a single location
>   * SNO1 and SNO2 are grouped in the same subnet
>   * SNO1 has 5 nodes
>   * SNO2 has 1 node
>   * By the grouping SNO1 will store more data because they will end up receiving 5 out of every 6 pieces sent to the group.

It is true that SNO1 would get more data than SNO2, and that is unfair, but I don't think it is true that this would cause the inverse effect of more concentrated data in a single location. Let us take the amount of data that would have gone to (SNO1+SNO2) before any IP subnet filtering, and call that amount **d**. SNO1 gets **d5/6** and SNO2 gets **d/6**. When IP subnet filtering is enabled, (SNO1+SNO2) together now receive **d/6** [1], because the satellite will consider all 6 nodes to be in the same location rather than being 6 separate locations. So SNO1 ends up with **5d/36**, which is less than **5d/6**. Ideally we would like for SNO1 and SNO2 to receive the same amount of data, for the sake of fairness, but the overall Storj network durability is not harmed. If this turns out to be a fairly common problem, we will of course need to rework the rules.

[1] The actual value will be somewhat higher, because the network will presumably also be considering other same-subnet nodes elsewhere as being in a common location, so some extra **O(1/n)** share of data goes to everyone else. But if those other same-subnet nodes really were in the same physical location, this is a desirable and correct outcome, so I'm not counting that increase in the calculation here.

> * It gives large data centers an advantage.
>   * By having multiple carriers available in a facility, with numerous discontinuous IP blocks available, a data center could end up having a majority of stored customer data. Especially as these nodes would be higher performing with more bandwidth available.

Yes, there are certainly many ways in which SNOs could work around the IP subnet filtering. The intention is simply to make it a little bit more difficult to do so. There are further steps we can take; for example, identifying colocated nodes by seeing if they tend to go offline and back online at the same times.

> If the filter is built such that you are actually looking at the ISP IP block allocations, or advertised prefixes :

We are of full agreement that this sort of approach would have many problems.

> How do you solve the distribution problem, without encountering these issues? In short, you don't. In long:
> * Stop over-engineering solutions to try to prevent data clustering. Bluntly, your methods for doing so are flawed. At best, they can exclude large geographic areas from normal selection, at worst actually lead to increased centralization.

They are only flawed if the requirement is to prevent all data clustering entirely. Our approach here is only to lessen data clustering. Even a partial solution is worthwhile, as long as the downsides are not too heavy.

> * Keep IP grouping simple, a single IP can include N nodes, but have data distributed as if they were a single node.

Can you clarify what you mean by this? Are you saying that we should support N nodes sharing a single IP, or that we should treat nodes sharing an IP as the same node, or something else? We are very interested in better ways to avoid data clustering.

> * Let the already described reputation system handle SNOs that are trying to game the system. If they are overtaxing their available resources they will be naturally limited as a result of having under-performing nodes.

We have encountered SNOs trying to game the system by making many node identities, all of which store data on the same SAN or in the same DC cabinet or the like. They aren't overtaxing their available resources, at all, they are just getting a larger share of data than they should be getting. The downside to the network is not just datashare, though; reliability suffers because multiple related pieces may be stored in the same place and be destroyed or taken offline together. This is the type of thing we would like to make harder.

> * In the node selection process introduce randomness. Randomly select nodes that fit the customer's requirements (Geofenced, reputation, etc.), and not just the best.

This already happens! The problem is just that it is still too likely for multiple related pieces to end up on related nodes.

> * Accept that true, enforced decentralization, is not easily code-able, but rather involves human curation and verification. If customers want this type of service, offer it as a tier to offset the work required to verify and track SNOs. Pay verified SNOs a premium to store this enforced data. Build a database of actual locations that can be queried and served as requested.

We do of course recognize that we can't 'enforce' decentralization 100%. The goal is simply to reduce data clustering until the reliability of data falls within the required parameters.

> Final note: I request that someone from the core Storj Labs team to actually read this. Don't just paste generic responses boiling down to "decentralization at all costs". Don't just close it saying "We're working on it." If there isn't an answer, leave it open until there is. If you can't publicly disclose this information, that's fine, just give some reassurance that this has actually been thought about and resolved.

Request granted! Please do let us know if you have further thoughts.

**Comment by JL421** [comment]
I very much appreciate the detailed response.

> Certainly accurate. By our experience across many common ISPs worldwide, however, it is a reasonably good approximator.

That model may hold up very well in dense city environments, but I'm regularly work with ISPs who spread a /24 over an area a bit larger than the city of Paris (130 sq km.). That's even averaging all of their holdings serving communities between 100 and 15,000 people, so even that density varies drastically.

> I was not aware of LISP. Thanks for the pointer! It looks like LISP adoption would actually make this approach more effective, as one of the guiding principles is that "RLOCs must be allocated in a way that is congruent with the topology of the network". Thus sites with a common RLOC (or possibly we'd look at a common RLOC prefix) will be more likely to be located in the same physical space, as they are guaranteed to be in the same topological spot on the network. Have I got that wrong? 
> ...
> Certainly true. It does not appear, though, that any ISPs require all traffic to be encapsulated with LISP, and the Storj network is unlikely to make use of LISP encapsulation either (since it establishes endpoint identity in a different way), so LISP will probably not be a major factor here.

That's a fairly good understanding of LISP, but there are a few additional points. The "RLOCs must be allocated in a way that is congruent with the topology of the network" is actually talking about the current addressing scheme. LISP actually breaks Rekhter's Law and makes addressing not follow topology. The real limitation is router interconnect speed, router memory, and latency. The general rule is so long as you are within 15 milliseconds of your LISP RLOC database, and you have the bandwidth available deal with possibly sending the traffic down a wire twice, you're golden. Light in a fiber travels pretty far in 15ms, and if I'm an ISP I can take the 8 /24s I used to allocate to a rural CO, consolidate them down to 4 /24s for the same large area, I will.

As for the encapsulation aspect, that's only once it has hit the provider's Internet edge. At that point, it is sent to the LISP Control Plane, encapsulated and sent to its destination router, unencapsulated, and finally forwarded to its endpoint.

> It is true that SNO1 would get more data than SNO2, and that is unfair, but I don't think it is true that this would cause the inverse effect of more concentrated data in a single location. Let us take the amount of data that would have gone to (SNO1+SNO2) before any IP subnet filtering, and call that amount d. SNO1 gets d5/6 and SNO2 gets d/6. When IP subnet filtering is enabled, (SNO1+SNO2) together now receive d/6 [1], because the satellite will consider all 6 nodes to be in the same location rather than being 6 separate locations. So SNO1 ends up with 5d/36, which is less than 5d/6.

So that response makes it seem like you consider a geographic region covered by the subnet to be a "single location", and this actually agrees with my argument. If the goal is to treat the subnet as a large failure domain with data spread evenly throughout, SNOs with multiple nodes will skew that distribution. Those SNOs will concentrate data into their single location (building). Since we are still counting the entire area as one node for distribution, that may or may not still fit your goal, but within that distribution we are reducing the failure domain.

> Can you clarify what you mean by this? Are you saying that we should support N nodes sharing a single IP, or that we should treat nodes sharing an IP as the same node, or something else? We are very interested in better ways to avoid data clustering.

Yes, I am saying shrink the scope from treating an entire subnet as one node for distribution, to a single IP. When the IP filter discussion came up that is how it was initially described, but expanding the scope to include up to 253 endpoints in a single group seems a bit much. Granted I never anticipate that kind of density unless someone is actually running a data center.

> We have encountered SNOs trying to game the system by making many node identities, all of which store data on the same SAN or in the same DC cabinet or the like. They aren't overtaxing their available resources, at all, they are just getting a larger share of data than they should be getting. The downside to the network is not just datashare, though; reliability suffers because multiple related pieces may be stored in the same place and be destroyed or taken offline together. This is the type of thing we would like to make harder.

Yes but if we look at the entire network the probability of that actually happening should be less than the network durability SLA based on node count. There is an upper limit to what you can put in a single physical space, and you will be limited by either IO, CPU, or Bandwidth at a point. The SNOs that can actually handle the load requirement and not have their reputation suffer should still be a vast minority of the network, especially as these people are trying to maximize profit. The cost of hardware and bandwidth will be its own deterrent to that kind of behavior. Worst case where these are bad actors, so long as a coordinated group doesn't exceed the repair threshold of network control, there still isn't a major issue apart from possible repair costs, which are take care of already. If that theoretical group has enough compute to control more than that threshold, filtering by /24s and /64s wasn't going to be a hindrance anyway.

> This already happens! The problem is just that it is still too likely for multiple related pieces to end up on related nodes.

Like I mentioned above, if a cluster of nodes can actually keep its reputation up enough to store multiple related pieces, what's the probability that number actually crosses the repair threshold?

-----

At the end of the day, I'm alright with IP filtering, for a single IP. If you want to start grouping larger clusters into the same failure domain for data resiliency purposes, you need to look at manual verification for SNOs and their nodes, or heuristics like which clusters go offline at the same time. Unless you have data that shows otherwise, I'm not convinced that someone, or the geographic region that would cover a /24 could feasibly affect overall network resiliency to a degree where it would fail the SLA (remembering that 99.999% is still 1 unrecoverable file per 100,000). If it would, the scale of an event that would cause that would seem larger and more disruptive than losing the Tardigrade network. Such an event would seem significantly less likely than the satellites themselves failing and losing data associated with them.

Thank you for the response, it was very informative.

**Comment by thepaul** [comment]
> I very much appreciate the detailed response.

Likewise! I want to thank you again for engaging with us; this is already an informative and productive discussion from my end.

> That model may hold up very well in dense city environments, but I'm regularly work with ISPs who spread a /24 over an area a bit larger than the city of Paris (130 sq km.). That's even averaging all of their holdings serving communities between 100 and 15,000 people, so even that density varies drastically.

Hrm, that is new information to me. To clarify, though, these ISPs are still allocating one /32 address per customer? Or are they doing CGN or some other sort of translation in order to share addresses between multiple customers?

That does sound like there will be a higher possibility than expected of storage nodes ending up on the same /24 by accident.

> That's a fairly good understanding of LISP, but there are a few additional points. The "RLOCs must be allocated in a way that is congruent with the topology of the network" is actually talking about the current addressing scheme. LISP actually breaks Rekhter's Law and makes addressing not follow topology. The real limitation is router interconnect speed, router memory, and latency. The general rule is so long as you are within 15 milliseconds of your LISP RLOC database, and you have the bandwidth available deal with possibly sending the traffic down a wire twice, you're golden. Light in a fiber travels pretty far in 15ms, and if I'm an ISP I can take the 8 /24s I used to allocate to a rural CO, consolidate them down to 4 /24s for the same large area, I will.
>
> As for the encapsulation aspect, that's only once it has hit the provider's Internet edge. At that point, it is sent to the LISP Control Plane, encapsulated and sent to its destination router, unencapsulated, and finally forwarded to its endpoint.

I see! So LISP needs to be implemented at the ISP level, not by encapsulating on the consumer side. I suppose the name should have tipped me off, but I missed that detail. Thank you for the clarification. Do you have an idea of how common LISP deployments are currently?

> So that response makes it seem like you consider a geographic region covered by the subnet to be a "single location", and this actually agrees with my argument. If the goal is to treat the subnet as a large failure domain with data spread evenly throughout, SNOs with multiple nodes will skew that distribution. Those SNOs will concentrate data into their single location (building). Since we are still counting the entire area as one node for distribution, that may or may not still fit your goal, but within that distribution we are reducing the failure domain.

Ah, now I understand your original framing. You are right- if our determination of the failure domain is significantly wrong, it could hurt our durability enough to offset any gains we get from the crude clustering avoidance strategy.

> Yes, I am saying shrink the scope from treating an entire subnet as one node for distribution, to a single IP. When the IP filter discussion came up that is how it was initially described, but expanding the scope to include up to 253 endpoints in a single group seems a bit much. Granted I never anticipate that kind of density unless someone is actually running a data center.

Thanks for the clarification.

> Yes but if we look at the entire network the probability of that actually happening should be less than the network durability SLA based on node count. There is an upper limit to what you can put in a single physical space, and you will be limited by either IO, CPU, or Bandwidth at a point. The SNOs that can actually handle the load requirement and not have their reputation suffer should still be a vast minority of the network, especially as these people are trying to maximize profit. The cost of hardware and bandwidth will be its own deterrent to that kind of behavior. Worst case where these are bad actors, so long as a coordinated group doesn't exceed the repair threshold of network control, there still isn't a major issue apart from possible repair costs, which are take care of already. If that theoretical group has enough compute to control more than that threshold, filtering by /24s and /64s wasn't going to be a hindrance anyway.

Our data science team has, if I understand correctly, determined that the probability of having multiple pieces end up in the same failure-domain-cluster is too high, if we allow such things to be made trivially. From our experience with the v2 network and with the input from our resident "evil geniuses" (we hired a couple of the biggest abusers of the v2 system :smile:), we believe that the costs of hardware and bandwidth will not be enough of a deterrent. Node operators will go to very impressive (to me) lengths to earn more money or cheat the system.

I don't have the math or the model in front of me, and it's too much for me to figure in my head, but I suspect there is some Birthday Problem type action that shows up in the calculations for the probability of multiple pieces getting clustered into a common failure domain. The probability of any one given segment having multiple pieces end up in an EvilCo cluster is quite low, but the probability of _any_ segment of _any_ object having multiple pieces on _any_ cooperating cluster over an expected
average object lifetime is too high.

> Like I mentioned above, if a cluster of nodes can actually keep its reputation up enough to store multiple related pieces, what's the probability that number actually crosses the repair threshold?

I think it will be _easier_ to meet reputation requirements on a cluster of related nodes than it would be to maintain reputation on a single node with the same total amount of available bandwidth and disk. Each individual node will have its own downtime allotment, instead of a single downtime allotment for the one large node. Each node having less resources ought to mean less chance of a hardware or operational failure on any individual node. But we need to strongly encourage the one-large-node approach instead for the health of our data.

> At the end of the day, I'm alright with IP filtering, for a single IP. If you want to start grouping larger clusters into the same failure domain for data resiliency purposes, you need to look at manual verification for SNOs and their nodes, or heuristics like which clusters go offline at the same time. Unless you have data that shows otherwise, I'm not convinced that someone, or the geographic region that would cover a /24 could feasibly affect overall network resiliency to a degree where it would fail the SLA (remembering that 99.999% is still 1 unrecoverable file per 100,000). If it would, the scale of an event that would cause that would seem larger and more disruptive than losing the Tardigrade network. Such an event would seem significantly less likely than the satellites themselves failing and losing data associated with them.

Ah, here is one potential point of disconnect. Our expected level of object durability for the Tardigrade network is not 99.999%, but 99.99999% (or even more for certain storage classes with higher Reed-Solomon parameters). As a network engineer, I'm sure you don't need me to illustrate how much of a difference those few extra nines make.

Even with those numbers, you are probably correct that one geographic region covered by a /24 is very unlikely to affect network resiliency very drastically. It is only when we consider that many SNOs may take the cluster approach that the problem shows up. Thus our efforts simply to discourage such clustering, and not to eliminate them entirely.

With what you've told me, though, we may need to rethink this crude approach to limiting. Possibly this might need to wait until we can gather more data about how storage nodes tend to behave and how much they tend to cluster naturally. We expect to be able to differentiate purposeful clustering from nodes accidentally showing up on the same /24 subnet, to some useful degree of accuracy, by correlating whether they tend to become unreachable at the same times. At some point, our clustering avoidance strategy should include such analysis automatically, but that is currently fairly far down the roadmap.

Thank you again for your input.

**Comment by Th3Van** [comment]
_@thepaul - "We are very interested in better ways to avoid data clustering."_

How about using ASN to share file pieces? _https://en.wikipedia.org/wiki/Autonomous_system_(Internet)_
Looking from a customers p.o.v. (and RS), it would be better to place pieces on as many different "networks" as possible, and not on a bunch of SN using the same ISP/DC network, even though they are all on different /24 IPv4 subnets.

Currently we have **447** uniq ASN's in use on the storj v3 network, shared among **1452 (ish)** ID's - a few of those ID's are not SN, (satellites/bootstraps etc.), but anyways :)
When a piece has to be uploaded, the satellite selects 130 random ASN's (all uniq), where as one SN are selected in each ASN  based on the /24 subnet filtering, or some other IP/SN filtering :)


**Top 25 - ID's vs. ASN**
*List generated using a simple bash script pulling JSON-data from RIPE's API : *https://stat.ripe.net/docs/data_api**
```
root@server002:/disk002/server031.backup# ./AS-finder-SNO2.sh    
-----------------------------------------------------------------------------------------------------
Report generated     :  Sun Jul 21 06:09:33 CEST 2019
Total # of ID(s)     :  1452
Total # of uniq ASNs :  447
-----------------------------------------------------------------------------------------------------
209 ID's from ASN 24940 - "HETZNER-AS - Hetzner Online GmbH"
86 ID's from ASN 16276 - "OVH - OVH SAS"
39 ID's from ASN 7922 - "COMCAST-7922 - Comcast Cable Communications"
38 ID's from ASN 4134 - "CHINANET-BACKBONE No.31"
36 ID's from ASN 6830 - "LGI-UPC - Liberty Global B.V."
31 ID's from ASN 7018 - "ATT-INTERNET4 - AT&T Services"
28 ID's from ASN 3320 - "DTAG - Deutsche Telekom AG"
27 ID's from ASN 701 - "UUNET - MCI Communications Services"
23 ID's from ASN 12876 - "AS12876 - ONLINE S.A.S."
21 ID's from ASN 8708 - "RCS-RDS - RCS & RDS SA"
16 ID's from ASN 51167 - "CONTABO - Contabo GmbH"
15 ID's from ASN 5089 - "NTL - Virgin Media Limited"
14 ID's from ASN 3215 - "AS3215 - Orange S.A."
13 ID's from ASN 8764 - "TELIA-LIETUVA - Telia Lietuva, AB"
13 ID's from ASN 2119 - "TELENOR-NEXTEL - Telenor Norge AS"
13 ID's from ASN 1136 - "KPN - KPN B.V."
11 ID's from ASN 577 - "BACOM - Bell Canada"
11 ID's from ASN 3249 - "ESTPAK - Telia Eesti AS"
11 ID's from ASN 31334 - "KABELDEUTSCHLAND-AS - Vodafone Kabel Deutschland GmbH"
11 ID's from ASN 209 - "CENTURYLINK-US-LEGACY-QWEST - CenturyLink Communications"
10 ID's from ASN 852 - "ASN852 - TELUS Communications Inc."
10 ID's from ASN 8473 - "BAHNHOF - Bahnhof AB"
10 ID's from ASN 42775 - "DSRT-NET - ADWORLDS TECHNOLOGIES LP"
10 ID's from ASN 3352 - "TELEFONICA_DE_ESPANA - TELEFONICA DE ESPANA"
10 ID's from ASN 15169 - "GOOGLE - Google LLC"
..snip..
```
Complete list with ID's : *http://server250.storj.dk/storj-v3/stats/id-by-asn/ID-by-ASN_Th3Van-201907210506.txt*

**Comment by thepaul** [comment]
Nice analysis!

I'm surprised that the IDs are that well distributed. I'd expected more homogeneity, honestly. This could be a very good approach to avoiding clustering. More research is probably needed. If we just say "each segment can have at most one piece per ASN", how drastically would that limit the amount of data going to SNOs on Hetzner, for example? And how much would that upset our existing pioneer SNOs?

**Comment by phutchins** [comment]
I don't know that we'd want to purely limit by ASN in all cases. For example, Comcast doesn't all go down at once (https://istheservicedown.com/problems/comcast-xfinity/map). Definitely seems like something to factor in tho.

**Comment by stefanbenten** [comment]
@phutchins Comcast has multiple ASN numbers from what i know.

**Comment by iglesiasbrandon** [comment]
We now avoid clustering using several additional differentiators, not just IP filtering. We are going to close this issue.
